### PR TITLE
amp-list: customize array location in the response

### LINKED ISSUE #6507: Allow specifying the name of the `items` property for amp-list responses
Currently, the response from a CORS endpoint supplying data for amp-list [must be a JSON object that contains an array property "items"](https://www.ampproject.org/docs/reference/components/amp-list#usage).

An unfortunate consequence of this is that a large number of public APIs are unusable, because they dont' return data wrapped in an `items` array. For [example](https://api.github.com/users/dandv/repos), GitHub's API returns directly an array of objects.

Here's a small `amp-list` sample trying to show a user's repos: https://jsbin.com/xekavu/edit?html,output

Proposal: add an attribute to `amp-list` allowing the developer to specify the name of the items property (default "items"):

```html
<amp-list 
    layout=fixed-height
    items=""    
    src="https://api.github.com/users/dandv/repos">
```

### PR DESCRIPTION
**Pull request by dvoytenko**
Fixes #6507.

/cc @dandv, @ericlindley-g for the feature
/cc @Gregable @honeybadgerdontcare @powdercloud for validator changes
/cc @bpaduch for spec/docs changes

### CODE REVIEW

**Comment by Gregable** [comment]
looks good for validator changes. ping me when you've submitted it.

File: src/json.js [hunk 1]
```diff
@@ -82,6 +82,11 @@ export function recreateNonProtoObject(obj) {
  * @return {?JSONValueDef|undefined}
  */
 export function getValueForExpr(obj, expr) {
+  // The `.` indicates "the object itself".
+  if (expr == '.') {
+    return obj;
+  }
+  // Otherwise, navigate via properties.
```
Comment (cramforce) [inline src/json.js]: Switch `hasOwnProperty` to `Object.prototype.hasOwnProperty.call`

Comment (dvoytenko) [inline src/json.js]: Done

**Comment by ericlindley-g** [comment]
Nice—excited to get this enhancement out!

**Comment by dandv** [comment]
Thanks! When should we see this in production?

**Comment by dvoytenko** [comment]
After Jan 5th. Please see https://github.com/ampproject/amphtml/releases

**Comment by jalajc** [comment]
Oh, so still slated for release... 
As the documentation for this change is already updated [here](https://www.ampproject.org/docs/reference/components/amp-list), I was trying... to no avail and getting error ```Response must be {items: []} object```. Would appretiate is documentation includes a nested navigation example (or may be at https://www.ampbyexample.com).

(feature request is somewhat referenced here too: #3554 )

**Comment by dvoytenko** [comment]
@jalajc This feature should be already on PROD (via #6745). Do you have an example you tried with? It should work, so if not - please send it to me and I'll debug.

**Comment by jalajc** [comment]
@dvoytenko, Yups it was in prod the next day I wrote here. Nested list is perfect now.

 I didn't update here as I was facing some issue with cliping of content (in terms of visibility on ui) which was not happening earlier and I was trying to figure out if my config/html layout is faulty.

**Comment by jalajc** [comment]
One thing I notice is for nested list item, you still have to specify the src mandatorily and even src url is same that url endpoint is called as many times as you specify it.

To elaborate consider the following snippet for which mustache template is already defined as `relatedLy`
```
<amp-list width=300 height=75 layout=responsive template="relatedLy" src='https://suraurgeet_com/json/testrel.json' items="items1"></amp-list>
<amp-list width=300 height=75 layout=responsive template="relatedLy" src='https://suraurgeet_com/json/testrel.json' items="items2"></amp-list>
<amp-list width=300 height=75 layout=responsive template="relatedLy" src='https://suraurgeet_com/json/testrel.json' items="items3.items4"></amp-list>
```
Notice that 
1. **src** attribute is **same** in all three and its response contains objects items1/2/3/4
2. **items** attribute is **different** in all three

Now whats happening is `src`, the CORS endpoint, is called three times!!

Can some configuration be done to avoid unnecessary repeated call to CORS end point.
Can `src `be not made optional, in that case use url specified in template and issue a warning in console in dev mode; but call that end point only once in a session as its not a live-list.

(Its like I am billed 3x for api that actually owe 1x 😃, plus more processing and latency )

**Comment by dvoytenko** [comment]
@cramforce Would this be solved by the `datasource` idea that you had?

**Comment by jridgewell** [comment]
This will be handled by [#7562](https://github.com/ampproject/amphtml/pull/7562/files#diff-cb5b59ffc6c4d1ed0c99e9f2b5d748b6R68).

**Comment by cramforce** [comment]
Yes #7562 will do this.

**Pull request merged**